subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingTestOperationListener.java

```diff
@@ -74,8 +82,21 @@ public void started(BuildOperationDescriptor buildOperation, OperationStartEvent
         }
     }
 
+    private int idx = 1;
+
     @Override
     public void progress(OperationIdentifier buildOperationId, OperationProgressEvent progressEvent) {
+        if (progressEvent.getDetails() instanceof TestListenerBuildOperationAdapter.OutputProgress) {
+            TestListenerBuildOperationAdapter.OutputProgress progress = (TestListenerBuildOperationAdapter.OutputProgress) progressEvent.getDetails();
+            TestOutputEvent event = progress.getOutput();
+
+            InternalOperationDescriptor descriptor = new DefaultTestOutputDescriptor(buildOperationId + "." + idx++, "todo", "todo", progress.getTestId());
+            eventConsumer.started(new DefaultTestOutputStartedProgressEvent(progressEvent.getTime(), descriptor));
+
+            int destination = progress.getOutput().getDestination() == TestOutputEvent.Destination.StdErr ? Destination.StdErr.getCode() : Destination.StdOut.getCode();
+            DefaultTestOutputResult result = new DefaultTestOutputResult(progressEvent.getTime(), progressEvent.getTime(), destination , progress.getOutput().getMessage());
+            eventConsumer.finished(new DefaultTestOutputFinishedProgressEvent(progressEvent.getTime(), descriptor, result));
```

Comment:
Do we need separate started/finished events for these and a new operation type?

Reply:
Not necessarily, I only tried to follow the existing patterns. There's no clear documentation on what is expected on the client-side AFAIC.

Maybe we could go with a similar approach to how the wrapper download progress is handled.

Reply:
I think a new `TestOutputEvent` should suffice. Have you already discussed this with @big-guy or @adammurdoch?

Reply:
Adam OoO, let's see what @big-guy has to say.